### PR TITLE
Templated `FactorGraph::at` method

### PR DESCRIPTION
**Pull request by varunagrawal**
I added a templated `at` method to the `FactorGraph` base class, so we can switch from doing
```cpp
auto factor = std::dynamic_pointer_cast<HessianFactor>(graph.at(0));
```
to
```cpp
auto factor = graph.at<HessianFactor>(0);
```
which IMHO looks much cleaner and is easier to type out.

I return the shared pointer by value following [this stackoverflow answer](https://stackoverflow.com/a/10643624/1236990).

### CODE REVIEW

**Comment by dellaert** [comment]
Why not to develop?

**Comment by varunagrawal** [comment]
I'm cascading PRs so I can get all these features to work together for my stuff (in contrast to having to switch branches and compile).

Having it target the parent branch makes reviewing easier for you. :)

**Comment by varunagrawal** [comment]
@dellaert this PR has become tricky to review, but the change is exactly what the PR comment says it is.

File: gtsam/hybrid/tests/testHybridGaussianFactorGraph.cpp [hunk 1]
```diff
@@ -613,11 +613,11 @@ TEST(HybridGaussianFactorGraph, assembleGraphTree) {
   // Create expected decision tree with two factor graphs:
 
   // Get mixture factor:
-  auto mixture = std::dynamic_pointer_cast<GaussianMixtureFactor>(fg.at(0));
+  auto mixture = fg.at<GaussianMixtureFactor>(0);
```
Comment (varunagrawal) [inline gtsam/hybrid/tests/testHybridGaussianFactorGraph.cpp:616]: This is an example of the usage working for a test.

File: gtsam/inference/FactorGraph.h [hunk 2]
```diff
@@ -310,6 +310,21 @@ class FactorGraph {
...
+  template <typename F>
+  std::shared_ptr<F> at(size_t i) {
+    return std::dynamic_pointer_cast<F>(factors_.at(i));
+  }
+
+  /// Const version of templated `at` method.
+  template <typename F>
+  const std::shared_ptr<F> at(size_t i) const {
+    return std::dynamic_pointer_cast<F>(factors_.at(i));
+  }
```
Comment (varunagrawal) [inline gtsam/inference/FactorGraph.h:326]: This part is the main contribution of this PR. Every other change that was originally part of this PR was an application of this new method.

**Review by dellaert — COMMENTED** [review]
So, again a lot of cmake changes in here that I don't fully understand are behavior changes. Can we focus the PR on the templated at method?

**Comment by varunagrawal** [comment]
I'll reopen those other PRs and undo the merge commits. That should resolve this.

**Comment by varunagrawal** [comment]
@dellaert updated

**Review by dellaert — APPROVED** [review]
Cool, nice PR now that all clutter is gone :-) Great new functionality, I like it a lot!

**Pull request merged**